File: app/src/main/java/com/dm/bomber/services/core/ParamsService.java
package com.dm.bomber.services.core;

import java.util.Objects;

import okhttp3.HttpUrl;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;

public abstract class ParamsService extends Service {

    protected String url;
    protected String method;

    protected Request.Builder request;
    protected HttpUrl.Builder builder;

    public ParamsService(String url, String method, int... countryCodes) {
        super(countryCodes);

        this.url = url;
        this.method = method;
    }

    public ParamsService(String url, int... countryCodes) {
        this(url, null, countryCodes);
    }

    public void run(OkHttpClient client, Callback callback, Phone phone) {
        request = new Request.Builder();
        builder = Objects.requireNonNull(HttpUrl.parse(url)).newBuilder();

        buildParams(phone);

        request.url(builder.build().toString());

        if (method != null)
            request.method(method, RequestBody.create("", null));

        client.newCall(request.build()).enqueue(callback);
    }

    public abstract void buildParams(Phone phone);
}

File: app/src/main/java/com/dm/bomber/services/core/ParamsService.kt
package com.dm.bomber.services.core

import okhttp3.HttpUrl
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.util.*

abstract class ParamsService(
    protected val url: String,
    protected val method: String?,
    vararg countryCodes: Int
) : Service(*countryCodes) {

    protected var request: Request.Builder = Request.Builder()

    protected lateinit var builder: HttpUrl.Builder

    constructor(url: String, vararg countryCodes: Int) : this(url, null, *countryCodes)

    override fun run(client: OkHttpClient, callback: Callback, phone: Phone) {
        builder = Objects.requireNonNull<HttpUrl>(url.toHttpUrlOrNull()).newBuilder()

        buildParams(phone)

        request.url(builder.build().toString())

        if (method != null)
            request.method(method, "".toRequestBody(null))

        client.newCall(request.build()).enqueue(callback)
    }

    abstract fun buildParams(phone: Phone)
}
